Retry Workspace creation to mitigate 503s

mitchnielsen opened the pull request:
Mitigates the 503 service unavailable errors by retrying Workspace creation. Further information provided in comments in the code for posterity.

Related to https://github.com/PrefectHQ/terraform-provider-prefect/issues/241

## Testing

Run the Acceptance Tests repeatedly:
- Old behavior: resource tests would sometimes fail to create Workspaces with a `503 Service Unavailable error`
- New behavior: resource tests pass every time

mitchnielsen committed:
Retry Workspace creation to mitigate 503s

Mitigates the 503 service unavailable errors by retrying Workspace
creation. Further information provided in comments in the code for
posterity.

Related to https://github.com/PrefectHQ/terraform-provider-prefect/issues/241

jimid27 reviewed (approved):
This Lgtm but I'm still no go expert

parkedwards commented on internal/provider/resources/workspace.go:
```diff
@@ -154,11 +156,25 @@ func (r *WorkspaceResource) Create(ctx context.Context, req resource.CreateReque
-	})
+	// In certain scenarios, such as in tests when multiple Workspaces are created at the same time
+	// for test isolation, Prefect will return a "503 Service Unavailable" error.
+	// To mitigate this, we will retry the Workspace creation. The Create method will be considered
+	// successful if either the Workspace was created or if the Workspace already exists.
```
```suggestion
	// To mitigate this, we will retry the Workspace creation. The Create method will stop retrying
	// if either the Workspace was created or if the Workspace already exists.
```

parkedwards commented on internal/client/workspaces.go:
```diff
@@ -57,7 +57,12 @@ func (c *WorkspacesClient) Create(ctx context.Context, data api.WorkspaceCreate)
+	// 2. StatusConflict (409) - meaning it already exists.
+	//   This is relevant in cases where the Create method is retried.
+	// If the error is not one of these, then Create is considered unsuccessful.
+	if resp.StatusCode != http.StatusCreated && resp.StatusCode != http.StatusConflict {
```
so if we get a 409, that means that the workspace exists already right

if we skip the early exit here on a 409, what's the expected behavior

parkedwards replied:
basically, im wondering if we should still exit on a 409 early, and find some way to pass signal back to the retry method to cease attempts (which i think is what we want to do)

parkedwards replied:
bc if we proceed here on a 409, the error will be some nil pointer exception from a failed json unmarshalling, i think

parkedwards replied:
maybe: https://github.com/avast/retry-go?tab=readme-ov-file#func--attemptsforerror

mitchnielsen replied:
I think I see what you mean - what you're suggesting would be more explicit and stable. Let me try 👍🏼

mitchnielsen replied:
@parkedwards - I started down that path but ran into some blockers/questions, so I pushed 834de5f34 to simplify this a bit. My reasoning is in the commit message - if you disagree I'm happy to try that option again assuming we can address the blockers/questions 👍🏼

parkedwards replied:
all good, i think the important part is retrying to begin with

mitchnielsen committed:
Rephrase comment

Co-authored-by: Edward Park <ed.sh.park@gmail.com>

mitchnielsen committed:
Ignore 409 status conflict on Workspace create

As part of retrying Create to work around 503 service unavailable
errors, we wanted to check for a 409 status conflict response because it
meant we didn't need to keep retrying the Create request.

However, when a 409 status conflict is reached, the Workspace is not
included in the response body. That means we'd have to run a separate
request to get the Workspace from the API - but the existing Get method
wants a workspace ID, which we don't have.

So we'd need to define a new GetByName method to get the method by the
name we originally wanted to use to create the workspace. But that
raises some questions:

- How do we handle a failure in that case? Do we add a retry at that
  point to avoid retrying the entire Create method and starting from
  square 1?
- Does it make sense to have a separate method just for this one error
  case?
- Would we ever even hit a 409 in the first place if we are sequentially
  running Create until the response is 201 created?
- We don't check for a 409 status conflict error on any other endpoints,
  is that a pattern we want to start following?

As an iteration, this commit simply removes the check for a 409 status
conflict error and relies on a simple retry mechanism. It will retry the
Create call until it returns 201 created.
